### PR TITLE
Add Validation When Add a Charged Token.

### PR DESCRIPTION
**Pull request by apweili**
1.  add that check the token if it is profitable and burnable in SetMethodFee and SetSymbolsToPayTxSizeFee.
2. check if the token weight may cause overflow in SetSymbolsToPayTxSizeFee.
3. add an api IsTokenAvailableForMethodFee to judge if the token can be set as fee.
4. modify unit tests.

### CODE REVIEW

File: contract/AElf.Contracts.Association/AssociationContract_ACS1_TransactionFeeProvider.cs [hunk 1]
```diff
@@ -90,7 +90,11 @@ private void AssertValidToken(string symbol, long amount)
 
             var tokenInfoInput = new GetTokenInfoInput {Symbol = symbol};
             var tokenInfo = State.TokenContract.GetTokenInfo.Call(tokenInfoInput);
-            Assert(tokenInfo != null && !string.IsNullOrEmpty(tokenInfo.Symbol), $"Token is not found. {symbol}");
+            Assert(tokenInfo != null && !string.IsNullOrEmpty(tokenInfo.Symbol),$"Token is not found. {symbol}");
+            var primaryTokenSymbol = (State.TokenContract.GetPrimaryTokenSymbol.Call(new Empty())).Value;
+            // ReSharper disable once PossibleNullReferenceException
+            if(primaryTokenSymbol != symbol)
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.Association/AssociationContract_ACS1_TransactionFeeProvider.cs:96]: Do not check primary symbol

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 2]
```diff
@@ -227,18 +227,18 @@ public override Empty SetSymbolsToPayTxSizeFee(SymbolListToPayTxSizeFee input)
...
-                    Assert(tokenInfo.AddedTokenWeight == 1 && tokenInfo.BaseTokenWeight == 1,
-                        $"symbol:{tokenInfo.TokenSymbol} weight should be 1");
+                    Assert(tokenWeightInfo.AddedTokenWeight == 1 && tokenWeightInfo.BaseTokenWeight == 1,
+                        $"symbol:{tokenWeightInfo.TokenSymbol} weight should be 1");
                 }
-
-                AssertSymbolToPayTxFeeIsValid(tokenInfo);
-                Assert(!symbolList.Contains(tokenInfo.TokenSymbol), $"symbol:{tokenInfo.TokenSymbol} repeat");
-                symbolList.Add(tokenInfo.TokenSymbol);
+                else
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs:238]: Remove else

File: contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs [hunk 3]
```diff
@@ -682,12 +682,15 @@ public override Address GetFeeReceiver(Empty input)
...
+        private void AssertSymbolToPayTxFeeIsValid(SymbolToPayTxSizeFee tokenWeightInfo)
         {
-            Assert(!string.IsNullOrEmpty(tokenInfo.TokenSymbol) & tokenInfo.TokenSymbol.All(IsValidSymbolChar),
-                "Invalid symbol.");
-            Assert(tokenInfo.AddedTokenWeight > 0 && tokenInfo.BaseTokenWeight > 0,
-                $"symbol:{tokenInfo.TokenSymbol} weight should be greater than 0");
+            Assert(tokenWeightInfo.AddedTokenWeight > 0 && tokenWeightInfo.BaseTokenWeight > 0,
+                $"symbol:{tokenWeightInfo.TokenSymbol} weight should be greater than 0");
+            var tokenInfo = State.TokenInfos[tokenWeightInfo.TokenSymbol];
+            Assert(tokenInfo != null && tokenInfo.IsProfitable, $"Token is not found. {tokenWeightInfo.TokenSymbol}");
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.MultiToken/TokenContract_Fees.cs:690]: tokenInfo.IsProfitable duplicate

File: contract/AElf.Contracts.Election/ElectionContract_ACS1_TransactionFeeProvider.cs [hunk 4]
```diff
@@ -91,7 +91,11 @@ private void AssertValidToken(string symbol, long amount)
 
             var tokenInfoInput = new GetTokenInfoInput {Symbol = symbol};
             var tokenInfo = State.TokenContract.GetTokenInfo.Call(tokenInfoInput);
-            Assert(tokenInfo != null && !string.IsNullOrEmpty(tokenInfo.Symbol), $"Token is not found. {symbol}");
+            Assert(tokenInfo != null && !string.IsNullOrEmpty(tokenInfo.Symbol),$"Token is not found. {symbol}");
+            var primaryTokenSymbol = (State.TokenContract.GetPrimaryTokenSymbol.Call(new Empty())).Value;
+            // ReSharper disable once PossibleNullReferenceException
+            if(primaryTokenSymbol != symbol)
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.Election/ElectionContract_ACS1_TransactionFeeProvider.cs:97]: Need to remove

File: contract/AElf.Contracts.Genesis/BasicContractZeroContract_ACS1_TransactionFeeProvider.cs [hunk 5]
```diff
@@ -75,7 +75,11 @@ private void AssertValidToken(string symbol, long amount)
 
             var tokenInfoInput = new GetTokenInfoInput {Symbol = symbol};
             var tokenInfo = State.TokenContract.GetTokenInfo.Call(tokenInfoInput);
-            Assert(tokenInfo != null && !string.IsNullOrEmpty(tokenInfo.Symbol), $"Token is not found. {symbol}");
+            Assert(tokenInfo != null && !string.IsNullOrEmpty(tokenInfo.Symbol),$"Token is not found. {symbol}");
+            var primaryTokenSymbol = (State.TokenContract.GetPrimaryTokenSymbol.Call(new Empty())).Value;
+            // ReSharper disable once PossibleNullReferenceException
+            if(primaryTokenSymbol != symbol)
```
Comment (jason-hoopox) [inline contract/AElf.Contracts.Genesis/BasicContractZeroContract_ACS1_TransactionFeeProvider.cs:81]: Need to remove

**Review by EanCuznaivy — APPROVED** [review]

**Review by jason-hoopox — APPROVED** [review]

**Review by shiwk — APPROVED** [review]

**Pull request merged**